support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -45,367 +27,121 @@ class CreateZuoraSubscription(servicesProvider: ServiceProvider = ServiceProvide
     services: Services
   ): FutureHandlerResult = {
 
-    val createSubscription = ZuoraSubscriptionCreator(
-      () => DateTime.now(DateTimeZone.UTC),
-      services.promotionService,
-      services.redemptionService,
-      services.zuoraService,
-      services.giftCodeGenerator,
-      services.config.zuoraConfigProvider.get(state.user.isTestUser),
-      state.user.isTestUser,
-    )
-
-    maybeDigitalSubscriptionGiftRedemption(state.product, state.paymentMethod) match {
-      case Some(redemptionData) =>
-        redeemGift(redemptionData, requestInfo, state, services.zuoraGiftService, services.catalogService)
-      case None =>
-        createSubscription.create(state, requestInfo)
+    val zuoraProductHandlers = new ZuoraProductHandlers(services, state)
```

Comment:
Seems a bit wasteful to construct 7 objects when we're only going to use one. Could you move the construction of the appropriate handler class into the match statement below?
Eg.
```scala
case redemptionState: DigitalSubscriptionGiftRedemptionState =>
        new ZuoraDigitalSubscriptionGiftRedemptionHandler(
          services.zuoraGiftService,
          services.catalogService,
          state.user,
          state.requestId,
        ).redeemGift(redemptionState)
```

Reply:
agree and I would like to improve that - it just seemed to obscure the elegance of the match statement, I will have a try again.

Reply:
this is still to do and should be reasonably self contained, all the other comments are resolved

Reply:
will change all to lazy val (after discussion)